### PR TITLE
feat: add transfer_manager.upload_chunks_concurrently using the XML MPU API

### PR DESCRIPTION
**Pull request by andrewsg**
This adds support for XML MPU multiprocess uploads to Transfer Manager.

### CODE REVIEW

File: google/cloud/storage/transfer_manager.py [hunk 1]
```diff
@@ -844,9 +855,208 @@ def download_chunks_concurrently(
...
+    # Pickle the blob ahead of time (just once, not once per chunk) if needed.
+    maybe_pickled_client = _pickle_client(client) if needs_pickling else client
+
+    futures = []
+
+    with pool_class(max_workers=max_workers) as executor:
+
+        for part_number in range(1, num_of_parts + 1):
+            start = (part_number - 1) * chunk_size
+            end = min(part_number * chunk_size, size)
```
Comment (cojenco) [inline google/cloud/storage/transfer_manager.py:993]: Could you remind me again if the byte ranges start/end are inclusive? 

I'm wondering for the intermediate parts, should this instead be `end = min(part_number * chunk_size - 1, size)`?

File: google/cloud/storage/transfer_manager.py [hunk 2]
```diff
@@ -844,9 +855,208 @@ def download_chunks_concurrently(
...
+
+    try:
+        # Harvest results and raise exceptions.
+        for future in futures:
+            part_number, etag = future.result()
+            container.register_part(part_number, etag)
+
+        container.finalize(blob._get_transport(client))
+    except Exception:
+        container.cancel(blob._get_transport(client))
```
Comment (cojenco) [inline google/cloud/storage/transfer_manager.py:1021]: Shall we also document that the MPU will be cancelled if not finalized completely

File: google/cloud/storage/transfer_manager.py [hunk 3]
```diff
@@ -844,9 +855,208 @@ def download_chunks_concurrently(
...
+        raise
+
+
+def _upload_part(
+    maybe_pickled_client, url, upload_id, filename, start, end, part_number, checksum
+):
+    """Helper function that runs inside a thread or subprocess to upload a part.
```
Comment (cojenco) [inline google/cloud/storage/transfer_manager.py:1028]: What are the plans on adding TM user metrics for MPU? We might want to expose extra_headers across MPU?

**Review by cojenco — COMMENTED** [review]
Looking good 🎉 have a few questions

File: google/cloud/storage/transfer_manager.py [hunk 4]
```diff
@@ -844,9 +855,208 @@ def download_chunks_concurrently(
...
+    hostname = _get_host_name(client._connection)
+    url = hostname + "/" + bucket.name + "/" + blob.name  # FIXME: make this nicer
+    content_type = blob._get_content_type(content_type, filename=filename)
+
+    container = XMLMPUContainer(url, filename)
+    container.initiate(transport=transport, content_type=content_type)
+    upload_id = container.upload_id
+
+    size = os.path.getsize(filename)
+    num_of_parts = -(size // -chunk_size)
```
[the remark this answers is not in the document]
Comment (andrewsg) [reply google/cloud/storage/transfer_manager.py:981]: Yep, this is a trick to do ceiling division without the math library. I'll comment.

File: google/cloud/storage/transfer_manager.py (hunk 1 above)
Comment (andrewsg) [reply google/cloud/storage/transfer_manager.py:993]: It's zero-indexed and exclusive, which means that if part 1 is 0-8 and part 2 is 8-16, then part 1 will have bytes 0, 1, 2, 3, 4, 5, 6, 7 and part 2 will have 8, 9, 10, 11, 12, 13, 14, 15. Both parts have zero bytes and the total is 16. Byte 16 doesn't exist. So this math should be accurate - I'll be sure to triple-check by actually using the method on some large files and checksumming them, though.

File: google/cloud/storage/transfer_manager.py (hunk 2 above)
Comment (andrewsg) [reply google/cloud/storage/transfer_manager.py:1021]: Will do.

File: google/cloud/storage/transfer_manager.py (hunk 3 above)
Comment (andrewsg) [reply google/cloud/storage/transfer_manager.py:1028]: This is private method, so no worries about changing it when we integrate user metrics, which I'll focus on next week.

**Comment by andrewsg** [comment]
Added integration tests. PTAL

File: tests/unit/test_transfer_manager.py [hunk 5]
```diff
@@ -600,6 +603,85 @@ def test_download_chunks_concurrently_passes_concurrency_options():
...
+                chunk_size=SIZE // 2,
+                worker_type=transfer_manager.THREAD,
+                max_workers=MAX_WORKERS,
+                deadline=DEADLINE,
+            )
+        except ValueError:
+            pass  # The futures don't actually work, so we expect this to abort.
+            # Conveniently, that gives us a chance to test the auto-delete
+            # exception handling feature.
+        container_mock.cancel.assert_called_once_with(transport)
```
[the remark this answers is not in the document]
Comment (andrewsg) [reply tests/unit/test_transfer_manager.py:684]: Really, testing exceptions is just breaking things on purpose

File: google/cloud/storage/transfer_manager.py [hunk 6]
```diff
@@ -844,9 +855,211 @@ def download_chunks_concurrently(
...
+    the documentation at https://cloud.google.com/storage/docs/multipart-uploads
+    before using this feature.
+
+    Downloads that fail due to an exception will be proactively canceled by the
+    library. If the download fails due to reason that precludes cancellation,
+    such as a hardware failure, process termination or power outage, then the
+    incomplete upload may persist indefinitely. To mitigate this, set
+    the `AbortIncompleteMultipartUpload` with a nonzero `Age` in bucket
+    lifecycle rules, or refer to the XML API documentation linked above to learn
+    more about how to list and delete individual downloads.
```
[the remark this answers is not in the document]
Comment (andrewsg) [reply google/cloud/storage/transfer_manager.py]: Yep, thanks!

File: google/cloud/storage/transfer_manager.py [hunk 7]
```diff
@@ -844,9 +855,211 @@ def download_chunks_concurrently(
...
+        The blob to which to upload.
+
+    :type content_type: str
+    :param content_type: (Optional) Type of content being uploaded.
+
+    :type chunk_size: int
+    :param chunk_size:
+        The size in bytes of each chunk to send. The optimal chunk size for
+        maximum throughput may vary depending on the exact network environment
+        and size of the blob.
```
[the remark this answers is not in the document]
Comment (andrewsg) [reply google/cloud/storage/transfer_manager.py]: Good call, added.

**Comment by andrewsg** [comment]
PTAL once more; this should be the final feature set I think

File: google/cloud/storage/transfer_manager.py [hunk 8]
```diff
@@ -844,9 +862,249 @@ def download_chunks_concurrently(
     return None
 
 
+def upload_chunks_concurrently(
+    filename,
+    blob,
+    content_type=None,
+    chunk_size=TM_DEFAULT_CHUNK_SIZE,
```
Comment (danielduhh) [inline google/cloud/storage/transfer_manager.py:869]: this is technically MPU part_size? Are using chunk_size for consistency?

File: google/cloud/storage/transfer_manager.py [hunk 9]
```diff
@@ -844,9 +862,249 @@ def download_chunks_concurrently(
...
+    """Upload a single file in chunks, concurrently.
+
+    This function uses the XML MPU API to initialize an upload and upload a
+    file in chunks, concurrently with a worker pool.
+
+    The XML MPU API is significantly different from other uploads; please review
+    the documentation at https://cloud.google.com/storage/docs/multipart-uploads
+    before using this feature.
+
+    The library will attempt to cancel uploads that fail due to an exception.
```
Comment (danielduhh) [inline google/cloud/storage/transfer_manager.py:886]: +1

File: google/cloud/storage/transfer_manager.py (hunk 8 above)
Comment (andrewsg) [reply google/cloud/storage/transfer_manager.py:869]: It is the part size. I am using chunk_size for consistency with the similar download method.

File: google/cloud/storage/transfer_manager.py [hunk 10]
```diff
@@ -1046,11 +1051,27 @@ def _upload_part(
...
     return (part_number, part.etag)
 
 
+def _headers_from_metadata(metadata):
+    """Helper function to translate object metadata into a header dictionary."""
+
+    headers = {}
+    # Handle standard writable metadata
+    for key, value in metadata.items():
+        if key in METADATA_HEADER_TRANSLATION:
```
Comment (cojenco) [inline google/cloud/storage/transfer_manager.py:1090]: QQ are there any writable metadata fields that don't require translation? 

Just wanted to make sure that they are all captured in _headers_from_metadata. If not, this works!

**Review by danielduhh — COMMENTED** [review]
Are you planning to add the `upload_sharded` command name for metrics on in an other PR?

**Comment by andrewsg** [comment]
@danielduhh Yes, actually that's already ready for review and I'll have it up as soon as this one is reviewed and merged.

**Review by cojenco — COMMENTED** [review]
LGTM, just have two questions

**Review by cojenco — APPROVED** [review]

File: google/cloud/storage/blob.py [hunk 11]
```diff
@@ -1697,7 +1697,7 @@ def _get_writable_metadata(self):
 
         return object_metadata
 
-    def _get_upload_arguments(self, client, content_type):
+    def _get_upload_arguments(self, client, content_type, filename=None):
```
[the remark this answers is not in the document]
Comment (andrewsg) [reply google/cloud/storage/blob.py:1700]: Yep, already resolved via merge in a different branch

File: google/cloud/storage/transfer_manager.py (hunk 10 above)
Comment (andrewsg) [reply google/cloud/storage/transfer_manager.py:1090]: I believe this is the complete list according to the documentation. Note that custom metadata and some other features like encryption are not supported through this translation dictionary but handled differently.

**Pull request merged**